gdal/ogr/ogrsf_frmts/pgeo/ogrpgeolayer.cpp

```diff
@@ -296,6 +296,27 @@ OGRFeature *OGRPGeoLayer::GetNextRawFeature()
 
         if( poGeom != nullptr && OGRERR_NONE == err )
         {
+            // always promote polygon/linestring geometries to multipolygon/multilinestring,
+            // so that the geometry types returned for the layer are predictable and match
+            // the advertised layer geometry type. See more details in OGRPGeoTableLayer::Initialize
+            const OGRwkbGeometryType eFlattenType = wkbFlatten(poGeom->getGeometryType());
+            if( eFlattenType == wkbPolygon )
```

Comment:
all below could likely be simplified to:
```suggestion
            if( eFlattenType == wkbPolygon || eFlattenType == wkbCurvePolygon || eFlatenType == wkbLineString || eFlattenType == wkbCompoundCurve )
            {
                poGeom = OGRGeometryFactory::forceTo(poGeom, OGR_GT_GetCollection(poGeom->getGeometryType()));
            }            
```

but regarding wkbCurvePolygon and wkbCompoundCurve, we should linearize them as a  MultiSurface/MultiCurve isn't a valid MultiPolygon/MultiLineString. The alternative would be to declare those generic wbkMultiSurface/wkbMultiCurve types as the layer geometry types. hum...

Reply:
For reference this logic was (mostly) copied from the openfilegdb driver, so any changes should also be applied there

Reply:
> For reference this logic was (mostly) copied from the openfilegdb driver, so any changes should also be applied there

I see... so yes, we have the potential flaw I underlined. I don't see a good solution to this. Linearizing is a pity, but advertizing the too-generic MultiSurface / MultiCurve could be an issue. I'd suggest to use my above proposed code simplification (with a comment about the issue with CurvePolygon/CompoundCurve), which hopefully should do the same as your code, and port it to the openfilegdb driver (where those CurvePolygon / CompoundCurve are tested as far as I can see)

Reply:
Actually it's not possible to use curved geometries in a personal geodatabase, so I'll clean up that logic